File: api/app/organizers/interactors/auth/login_user.rb
module Interactors::Auth
  class LoginUser < Interactors::BaseInteractor
    include Interactor::ContextValidator

    params :username, :password

    validate :username, :required
    validate :password, :required

    def call
      account = ::Auth::Account.new(
        type: ::Auth::AccountType::USER,
        username: username,
        password: password
      )

      session = Auth::Authentication.login(account)
      raise Error::ErrorCode.new('AUTHENTICATION_FAILED', :unauthorized) unless session

      context.data =  {
        token: session.access_token,
        expires_at: session.expires_at,
        user: session.user
      }
    end
  end
end

File: ui/src/graphql/mutations/auth/LoginUser.ts
import { gql } from 'graphql-tag'

export const LOGIN_USER_MUTATION = gql`
  mutation LoginUser($username: String!, $password: String!) {
    auth: loginUser(username: $username, password: $password) {
      token
      expiresAt
      user {
        id
        username
        name
      }
    }
  }
`
